File: python-prototype/simulation/crossing_state.py
from copy import deepcopy
from random import choice
from typing import Type

from simulation.config import Config

from .car import Car
from .constants import (
    STARTING_CARS_COUNTER,
)
from .crossing_connection.base.base_crossing_connection import (
    BaseCrossingConnection,
)
from .direction import Direction
from .type_hints import CarsCounterDict, LightsStateDict


class CrossingState:
    """Holds and handles crossing state."""

    lights_state_dict: LightsStateDict
    cars_counter_dict: CarsCounterDict
    crossing_connection: BaseCrossingConnection

    def __init__(
        self, CrossingConnectionImplementation: Type[BaseCrossingConnection]
    ) -> None:
        self.config = Config()

        self.crossing_connection = CrossingConnectionImplementation()
        self.lights_state_dict = self.crossing_connection.get_initial_lights_state()

        self.cars_counter_dict = deepcopy(STARTING_CARS_COUNTER)

    def __repr__(self) -> str:
        self.show_lights()

        return '\n'.join(
            [
                'Crossing state:',
                *[
                    '{}: {}🚗'.format(direction, car_number)
                    for direction, car_number in self.cars_counter_dict.items()
                ],
            ]
        )

    def show_lights(self) -> None:
        print('Lights 🚦')
        print(
            '\n'.join(
                [
                    *[
                        '{}:{}'.format(direction, light)
                        for direction, light in self.lights_state_dict.items()
                    ]
                ]
            )
        )

    def add_car(self, car: Car) -> None:
        self.cars_counter_dict[car.start] += 1

    def move_cars(self, cars_counter_dict: CarsCounterDict):
        self.crossing_connection.move_cars(
            lights_state_dict=self.lights_state_dict,
            cars_counter_dict=cars_counter_dict,
        )

    def check_if_there_are_any_cars(self) -> bool:
        return sum(self.cars_counter_dict.values()) == 0

    def add_random_car(self):
        return self.add_car(
            Car(start=choice(list(Direction)), destination=choice(list(Direction)))
        )

    def change_lights(self, winning_direction: Direction):
        """Changes intersection lights to green in direction where most cars are waiting."""
        self.lights_state_dict = self.crossing_connection.change_lights(
            lights_state_dict=self.lights_state_dict,
            winning_direction=winning_direction,
        )

File: intersection-api/src/crossingState.ts
import { CarsCounterRecord, LightsStateRecord } from './types';

import { BaseCrossingConnection } from './crossingConnection/base';
import Car from './car';
import Config from './config';
import Direction from './direction';
import { STARTING_CARS_COUNTER } from './constants';
import _ from 'lodash';

/**
 * Holds and handles crossing state.
 */
export default class CrossingState {
	lightsStateRecord: LightsStateRecord;
	carsCounterRecord: CarsCounterRecord;
	crossingConnection: BaseCrossingConnection;
	config: Config;

	constructor(
		CrossingConnectionImplementation: new () => BaseCrossingConnection
	) {
		(this.config = Config.instance),
			(this.crossingConnection = new CrossingConnectionImplementation()),
			(this.lightsStateRecord =
				this.crossingConnection.getInitialLightsState());
		this.carsCounterRecord = _.cloneDeep(STARTING_CARS_COUNTER);
	}

	toString(): string {
		this.showLights();
		return [
			'Crossing state:',
			...Object.entries(this.carsCounterRecord).map(
				([direction, carNumber]) => `${direction}: ${carNumber}🚗`
			),
		].join('\n');
	}

	showLights(): void {
		console.log('Lights 🚦');
		console.log(
			Object.entries(this.lightsStateRecord)
				.map(([direction, light]) => `${direction}:${light}`)
				.join('\n')
		);
	}

	addCar(car: Car): void {
		this.carsCounterRecord[car.start]++;
	}

	moveCars(carsCounterRecord: CarsCounterRecord): void {
		this.crossingConnection.moveCars(
			this.lightsStateRecord,
			carsCounterRecord
		);
	}

	checkIfThereAreAnyCars(): boolean {
		return (
			Object.values(this.carsCounterRecord).reduce(
				(acc, val) => acc + val,
				0
			) === 0
		);
	}

	addRandomCar(): void {
		this.addCar(
			new Car(
				_.sample(Object.values(Direction)) as Direction,
				_.sample(Object.values(Direction)) as Direction
			)
		);
	}

	changeLights(winningDirection: Direction): void {
		/**
		 * Changes intersection lights to green in direction where most cars are waiting.
		 */
		this.lightsStateRecord = this.crossingConnection.changeLights(
			(this.lightsStateRecord = this.lightsStateRecord),
			(winningDirection = winningDirection)
		);
	}
}
